core/app/models/spree/promotion_code/batch_builder.rb

```diff
@@ -39,13 +40,15 @@ def generate_random_codes
       new_codes = Array.new(max_codes_to_generate) { generate_random_code }.uniq
       codes_for_current_batch = get_unique_codes(new_codes)
 
-      codes_for_current_batch.each do |value|
+      codes_for_current_batch = codes_for_current_batch.map do |value|
         Spree::PromotionCode.create!(
           value: value,
           promotion: promotion,
           promotion_code_batch: promotion_code_batch
         )
-      end
+      rescue ActiveRecord::RecordInvalid
+        nil
+      end.compact
```

Comment:
To address @tvdeyen's concern, what about something like:

```ruby
codes_for_current_batch.filter! do |value|
  Spree::PromotionCode.create!(
    value: value,
    promotion: promotion,
    promotion_code_batch: promotion_code_batch
  )
rescue ActiveRecord::RecordInvalid
  false
end
```

This should maintain the existing performance characteristics.